File: scripts/shared/consolidated_tables_helper.py
"""
Consolidated Tables Helper - Python Version

Centraliza INSERTs para tabelas consolidadas:
- tech_trends (github, stackoverflow, npm, pypi)
- community_posts (hackernews, reddit, producthunt)
- patents (epo, wipo, uspto)
"""

import json
from typing import Optional, Dict, Any, List
from datetime import datetime
import psycopg2
from psycopg2.extras import RealDictCursor


class ConsolidatedTablesHelper:
    """Helper para gerenciar INSERTs em tabelas consolidadas"""
    
    def __init__(self, conn):
        """
        Args:
            conn: Conexão psycopg2
        """
        self.conn = conn
    
    def insert_tech_trend(
        self,
        source: str,
        name: str,
        category: Optional[str] = None,
        trend_type: Optional[str] = None,
        score: Optional[float] = None,
        rank: Optional[int] = None,
        stars: Optional[int] = None,
        forks: Optional[int] = None,
        views: Optional[int] = None,
        mentions: Optional[int] = None,
        growth_rate: Optional[float] = None,
        period_start: Optional[datetime] = None,
        period_end: Optional[datetime] = None,
        metadata: Optional[Dict[str, Any]] = None
    ):
        """Insert/Update tech trend"""
        
        query = """
        INSERT INTO sofia.tech_trends (
            source, name, category, trend_type,
            score, rank, stars, forks, views, mentions, growth_rate,
            period_start, period_end, metadata
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        ON CONFLICT (source, name, period_start) DO UPDATE SET
            score = EXCLUDED.score,
            rank = EXCLUDED.rank,
            stars = EXCLUDED.stars,
            forks = EXCLUDED.forks,
            views = EXCLUDED.views,
            mentions = EXCLUDED.mentions,
            growth_rate = EXCLUDED.growth_rate,
            metadata = EXCLUDED.metadata,
            collected_at = NOW()
        """
        
        with self.conn.cursor() as cur:
            cur.execute(query, (
                source, name, category, trend_type,
                score, rank, stars, forks, views, mentions, growth_rate,
                period_start, period_end,
                json.dumps(metadata) if metadata else None
            ))
        self.conn.commit()
    
    def insert_community_post(
        self,
        source: str,
        external_id: str,
        title: str,
        url: Optional[str] = None,
        content: Optional[str] = None,
        author: Optional[str] = None,
        score: Optional[int] = None,
        comments_count: Optional[int] = None,
        upvotes: Optional[int] = None,
        category: Optional[str] = None,
        tags: Optional[List[str]] = None,
        posted_at: Optional[datetime] = None,
        metadata: Optional[Dict[str, Any]] = None
    ):
        """Insert/Update community post"""
        
        query = """
        INSERT INTO sofia.community_posts (
            source, external_id, title, url, content,
            author, score, comments_count, upvotes,
            category, tags, posted_at, metadata
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        ON CONFLICT (source, external_id) DO UPDATE SET
            title = EXCLUDED.title,
            score = EXCLUDED.score,
            comments_count = EXCLUDED.comments_count,
            upvotes = EXCLUDED.upvotes,
            metadata = EXCLUDED.metadata,
            collected_at = NOW()
        """
        
        with self.conn.cursor() as cur:
            cur.execute(query, (
                source, external_id, title, url, content,
                author, score, comments_count, upvotes,
                category,
                json.dumps(tags) if tags else None,
                posted_at,
                json.dumps(metadata) if metadata else None
            ))
        self.conn.commit()
    
    def insert_patent(
        self,
        source: str,
        patent_number: str,
        title: Optional[str] = None,
        abstract: Optional[str] = None,
        applicant: Optional[str] = None,
        inventor: Optional[str] = None,
        ipc_classification: Optional[List[str]] = None,
        technology_field: Optional[str] = None,
        country_id: Optional[int] = None,
        applicant_country: Optional[str] = None,
        filing_date: Optional[datetime] = None,
        publication_date: Optional[datetime] = None,
        grant_date: Optional[datetime] = None,
        metadata: Optional[Dict[str, Any]] = None
    ):
        """Insert/Update patent"""
        
        query = """
        INSERT INTO sofia.patents (
            source, patent_number, title, abstract,
            applicant, inventor,
            ipc_classification, technology_field,
            country_id, applicant_country,
            filing_date, publication_date, grant_date,
            metadata
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        ON CONFLICT (source, patent_number) DO UPDATE SET
            title = EXCLUDED.title,
            abstract = EXCLUDED.abstract,
            metadata = EXCLUDED.metadata,
            collected_at = NOW()
        """
        
        with self.conn.cursor() as cur:
            cur.execute(query, (
                source, patent_number, title, abstract,
                applicant, inventor,
                ipc_classification, technology_field,
                country_id, applicant_country,
                filing_date, publication_date, grant_date,
                json.dumps(metadata) if metadata else None
            ))
        self.conn.commit()
    
    def batch_insert_tech_trends(self, trends: List[Dict[str, Any]]):
        """Batch insert tech trends"""
        for trend in trends:
            self.insert_tech_trend(**trend)
    
    def batch_insert_community_posts(self, posts: List[Dict[str, Any]]):
        """Batch insert community posts"""
        for post in posts:
            self.insert_community_post(**post)
    
    def batch_insert_patents(self, patents: List[Dict[str, Any]]):
        """Batch insert patents"""
        for patent in patents:
            self.insert_patent(**patent)

File: scripts/shared/consolidated-tables-helper.ts
/**
 * Consolidated Tables Helper
 * 
 * Centraliza INSERTs para tabelas consolidadas:
 * - tech_trends (github, stackoverflow, npm, pypi)
 * - community_posts (hackernews, reddit, producthunt)
 * - patents (epo, wipo, uspto)
 */

import { Pool, PoolClient } from 'pg';

// ============================================================================
// TYPES
// ============================================================================

export interface TechTrend {
    source: 'github' | 'stackoverflow' | 'npm' | 'pypi';
    name: string;
    category?: string;
    trend_type?: string;
    score?: number;
    rank?: number;
    stars?: number;
    forks?: number;
    views?: number;
    mentions?: number;
    growth_rate?: number;
    period_start?: Date;
    period_end?: Date;
    metadata?: Record<string, any>;
}

export interface CommunityPost {
    source: 'hackernews' | 'reddit' | 'producthunt' | 'devto';
    external_id: string;
    title: string;
    url?: string;
    content?: string;
    author?: string;
    score?: number;
    comments_count?: number;
    upvotes?: number;
    category?: string;
    tags?: string[];
    posted_at?: Date;
    metadata?: Record<string, any>;
}

export interface Patent {
    source: 'epo' | 'wipo' | 'uspto';
    patent_number: string;
    title?: string;
    abstract?: string;
    applicant?: string;
    inventor?: string;
    ipc_classification?: string[];
    technology_field?: string;
    country_id?: number;
    applicant_country?: string;
    filing_date?: Date;
    publication_date?: Date;
    grant_date?: Date;
    metadata?: Record<string, any>;
}

// ============================================================================
// CONSOLIDATED TABLES HELPER CLASS
// ============================================================================

export class ConsolidatedTablesHelper {
    private pool: Pool;

    constructor(pool: Pool) {
        this.pool = pool;
    }

    /**
     * Insert/Update tech trend
     */
    async insertTechTrend(trend: TechTrend, client?: PoolClient): Promise<void> {
        const db = client || this.pool;

        const query = `
      INSERT INTO sofia.tech_trends (
        source, name, category, trend_type,
        score, rank, stars, forks, views, mentions, growth_rate,
        period_start, period_end, metadata
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
      ON CONFLICT (source, name, period_start) DO UPDATE SET
        score = EXCLUDED.score,
        rank = EXCLUDED.rank,
        stars = EXCLUDED.stars,
        forks = EXCLUDED.forks,
        views = EXCLUDED.views,
        mentions = EXCLUDED.mentions,
        growth_rate = EXCLUDED.growth_rate,
        metadata = EXCLUDED.metadata,
        collected_at = NOW()
    `;

        await db.query(query, [
            trend.source,
            trend.name,
            trend.category,
            trend.trend_type,
            trend.score,
            trend.rank,
            trend.stars,
            trend.forks,
            trend.views,
            trend.mentions,
            trend.growth_rate,
            trend.period_start,
            trend.period_end,
            trend.metadata ? JSON.stringify(trend.metadata) : null,
        ]);
    }

    /**
     * Insert/Update community post
     */
    async insertCommunityPost(post: CommunityPost, client?: PoolClient): Promise<void> {
        const db = client || this.pool;

        const query = `
      INSERT INTO sofia.community_posts (
        source, external_id, title, url, content,
        author, score, comments_count, upvotes,
        category, tags, posted_at, metadata
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
      ON CONFLICT (source, external_id) DO UPDATE SET
        title = EXCLUDED.title,
        score = EXCLUDED.score,
        comments_count = EXCLUDED.comments_count,
        upvotes = EXCLUDED.upvotes,
        metadata = EXCLUDED.metadata,
        collected_at = NOW()
    `;

        await db.query(query, [
            post.source,
            post.external_id,
            post.title,
            post.url,
            post.content,
            post.author,
            post.score,
            post.comments_count,
            post.upvotes,
            post.category,
            post.tags ? JSON.stringify(post.tags) : null,
            post.posted_at,
            post.metadata ? JSON.stringify(post.metadata) : null,
        ]);
    }

    /**
     * Insert/Update patent
     */
    async insertPatent(patent: Patent, client?: PoolClient): Promise<void> {
        const db = client || this.pool;

        const query = `
      INSERT INTO sofia.patents (
        source, patent_number, title, abstract,
        applicant, inventor,
        ipc_classification, technology_field,
        country_id, applicant_country,
        filing_date, publication_date, grant_date,
        metadata
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
      ON CONFLICT (source, patent_number) DO UPDATE SET
        title = EXCLUDED.title,
        abstract = EXCLUDED.abstract,
        metadata = EXCLUDED.metadata,
        collected_at = NOW()
    `;

        await db.query(query, [
            patent.source,
            patent.patent_number,
            patent.title,
            patent.abstract,
            patent.applicant,
            patent.inventor,
            patent.ipc_classification,
            patent.technology_field,
            patent.country_id,
            patent.applicant_country,
            patent.filing_date,
            patent.publication_date,
            patent.grant_date,
            patent.metadata ? JSON.stringify(patent.metadata) : null,
        ]);
    }

    /**
     * Batch insert tech trends
     */
    async batchInsertTechTrends(trends: TechTrend[]): Promise<void> {
        const client = await this.pool.connect();
        try {
            await client.query('BEGIN');
            for (const trend of trends) {
                await this.insertTechTrend(trend, client);
            }
            await client.query('COMMIT');
        } catch (error) {
            await client.query('ROLLBACK');
            throw error;
        } finally {
            client.release();
        }
    }

    /**
     * Batch insert community posts
     */
    async batchInsertCommunityPosts(posts: CommunityPost[]): Promise<void> {
        const client = await this.pool.connect();
        try {
            await client.query('BEGIN');
            for (const post of posts) {
                await this.insertCommunityPost(post, client);
            }
            await client.query('COMMIT');
        } catch (error) {
            await client.query('ROLLBACK');
            throw error;
        } finally {
            client.release();
        }
    }

    /**
     * Batch insert patents
     */
    async batchInsertPatents(patents: Patent[]): Promise<void> {
        const client = await this.pool.connect();
        try {
            await client.query('BEGIN');
            for (const patent of patents) {
                await this.insertPatent(patent, client);
            }
            await client.query('COMMIT');
        } catch (error) {
            await client.query('ROLLBACK');
            throw error;
        } finally {
            client.release();
        }
    }
}

// ============================================================================
// CONVENIENCE FUNCTIONS
// ============================================================================

/**
 * Create helper instance
 */
export function createConsolidatedTablesHelper(pool: Pool): ConsolidatedTablesHelper {
    return new ConsolidatedTablesHelper(pool);
}
